typescript-deno/deno.jsonc

```diff
@@ -0,0 +1,13 @@
+{
+  "tasks": {
+    "start": "./script/start",
+    "test": "./script/test"
+  },
+  // https://github.com/guardian/csnx/blob/1be3fd98ed527ccf0f1b8e21f10039e23fdb3fb7/libs/%40guardian/tsconfig/tsconfig.json
+  "compilerOptions": {
+    "noImplicitReturns": true,
+    "noUncheckedIndexedAccess": true,
```

Comment:
I have a feeling this specific compiler option might create more friction than intended, as it is known to create false positives. The TypeScript development team have repeatedly resisted calls to make this rule a default due to its inflexibility; see [here](https://github.com/microsoft/TypeScript/issues/49169#issuecomment-1677712312). 

This appears to be an issue particularly when attempting to index into arrays in bounds-checked loops, see example in the [docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#checked-indexed-accesses---nouncheckedindexedaccess)
```typescript
function screamLines(strs: string[]) {
  // This will have issues
  for (let i = 0; i < strs.length; i++) {
    console.log(strs[i].toUpperCase());
   // Object is possibly 'undefined'.
  }
}
```
My experience doing coding exercise interviews is that candidates often reach for C-style for loops and this rule might end up creating more confusion than needed. I would therefore skip this rule if possible, but happy to hear your thoughts.

The other rules (`noImplicitReturns` and `noUnusedLocals`) seem like great additions.

Reply:
That’s a fair point… Do we think that this is an interesting point of discussion for interviews?

---

For any future candidate reading this, I’m sure you’ll make a good impression if you do this instead 😉:

```ts
function screamLines(strs: string[]) {
  for (const str of strs) {
    console.log(str.toUpperCase());
  }
}
```

Reply:
In the game of life exercise it's extremely common to use indexing into a 2d array to find the neighbours in a grid, and it raises interesting questions about how to do it safely. For accessing neighbours in a 2d array of arrays how would you do that whilst keeping this rule happy?

Reply:
@michaelclapham simply by checking whether the value is not `undefined`.

```ts
const grid = [
  [1, 1, 0],
  [0, 0, 0],
  [1, 0, 1],
];

const safe = grid[0]?.[0] // number | undefined
if (typeof safe === "number") {
  console.log("we know this is a safe lookup!", safe)
}

const unsafe = grid[9]?.[9] // number | undefined
if (typeof unsafe === "number") {
  console.log("This will never happen!", unsafe)
}
```

[Playground Link](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&target=9&ssl=16&ssc=1&pln=17&pc=1#code/MYewdgziA2CmB0w4EMBOAKAlAbgFC9EgBcACAc1QEsATEgXhIG1cSmBGAGhM5IAYBdDiya8uovoOGMe4tpP54C4CKQjIAZrHrkq1RgID88ff1wB6M6ysA9A7krqS6IgE8ADrBCO1m+nQYARGAArgC2AEawqAGYJADewoRQcPDQIGToAQDuWgDWYCBZJEQAFpQQJOUkyCQ+WmkgucFuAIQBXHWYuAC++EmkwZAaWgwUNIwAnPxGk6YWVqy29o7O7p6Og3V+gSERUTHxicowCGkZAQAqZRVZlNDQJGCwAG5RJCXIbh5gbVybw11ukA)